mesonbuild/interpreter/type_checking.py

```diff
@@ -448,3 +461,281 @@ def link_whole_validator(values: T.List[T.Union[StaticLibrary, CustomTarget, Cus
+    convertor=lambda x: None if isinstance(x, list) else x,
+)
+
+# A variant used internally for build targets, and is stricter than the public
+# one.
+_LINK_WITH_KW: KwargInfo[T.List[T.Union[BothLibraries, SharedLibrary, SharedModule, StaticLibrary, CustomTarget, CustomTargetIndex]]] = KwargInfo(
+    'link_with',
+    ContainerTypeInfo(
+        list,
+        (BothLibraries, SharedLibrary, StaticLibrary,
+         SharedModule, CustomTarget, CustomTargetIndex,
+         Dependency)),
+    default=[],
+    listify=True,
+    validator=lambda x: _link_with_error if isinstance(x, Dependency) else None,
+)
+
+_BUILD_TARGET_KWS: T.List[KwargInfo] = [
+    KwargInfo('build_by_default', bool, default=True, since='0.40.0'),
+    KwargInfo('build_rpath', str, default='', since='0.42.0'),
+    KwargInfo('d_debug', ContainerTypeInfo(list, (str, int)), default=[], listify=True),
+    INCLUDE_DIRECTORIES.evolve(name='d_import_dirs'),
+    D_MODULE_VERSIONS_KW,
+    KwargInfo('d_unittest', bool, default=False),
+    DEPENDENCIES_KW,
+    KwargInfo(
+        'extra_files',
+        ContainerTypeInfo(list, (str, File, CustomTarget, CustomTargetIndex)),
+        default=[],
+        listify=True,
+    ),
+    KwargInfo(
+        'gnu_symbol_visibility', str, default='', since='0.48.0',
+        validator=in_set_validator({'', 'default', 'internal', 'hidden', 'protected', 'inlineshidden'}),
+    ),
+    KwargInfo('implicit_include_directories', bool, default=True, since='0.42.0'),
+    INCLUDE_DIRECTORIES,
+    INSTALL_KW,
+    # TODO: Eventually we want this to just be the generic INSTALL_DIR_KW, but in the mean time
+    # this is allowed to be a `List[str | bool]` (not just `False`) so we have to handle it manually
+    KwargInfo(
+        'install_dir',
+        ContainerTypeInfo(list, (str, bool)),
+        default=[],
+        listify=True,
+    ),
+    INSTALL_MODE_KW.evolve(since='0.47.0'),
+    KwargInfo('install_rpath', str, default=''),
+    INSTALL_TAG_KW.evolve(since='0.60.0'),
+    KwargInfo('link_args', ContainerTypeInfo(list, str), default=[], listify=True),
+    KwargInfo(
+        'link_depends',
+        ContainerTypeInfo(list, (str, File, CustomTarget, CustomTargetIndex)),
+        default=[],
+        listify=True,
+    ),
+    KwargInfo(
+        'link_language',
+        (str, NoneType),
+        since='0.51.0',
+        validator=in_set_validator(set(compilers.all_languages)),
+    ),
+    LINK_WHOLE_KW.evolve(since='0.40.0'),
+    _NAME_PREFIX_KW,
+    _NAME_PREFIX_KW.evolve(name='name_suffix'),
+    NATIVE_KW,
+    KwargInfo(
+        'objects',
+        ContainerTypeInfo(list, (str, File, ExtractedObjects)),
+        default=[],
+        listify=True,
+    ),
+    OVERRIDE_OPTIONS_KW.evolve(since='0.40.0'),
+    KwargInfo(
+        'sources',
+        ContainerTypeInfo(list, (str, File, CustomTarget, CustomTargetIndex, GeneratedList)),
```

Comment:
JAR is actually a giant mess, because it accepts all kinds of arguments it absolutely shouldn't. `link_language`, 'link_whole`, `vala_*`, etc. Once this lands maybe @tristan957 would want to help clean up the Jar interface and deprecate all of the values we never should have accepted 😉

Reply:
I've added `StructuredSources` here. But the more I think about it, the more I think I need to just fix Jar right now... Sigh